proposals/code-signing-for-homebrew.md

```diff
@@ -0,0 +1,101 @@
+
+### Sigstore signatures for third-party taps
+
+While not in scope for this proposal, this work would also enable signing for third-party taps, such as those maintained by independent open source organizations or companies.
+
+For third-party taps hosted on GitHub, the signing flow would be identical: any bottle-building workflow (including reusing the official one) would sign for bottles.
+
+## Client-side signature verification in Homebrew
+
+The Sigstore signatures described above are submitted to each tap's object storage so that they can be verified by clients during the normal `brew install` flow.
+
+To verify a Sigstore signature, a particular `brew` client needs the following:
+
+- A trust proposition, i.e. a reason to trust the identity producing the signature. For Homebrew, this trust proposition comes from the tap's own identity: the act of performing `brew tap foo/bar` establishes that Sigstore identities tied to `github.com/foo/bar` can be trusted to sign for bottles in that tap. This applies to implicit `brew tap` invocations as well, e.g. `brew install foo/bar/baz` asserts that the user expects a signature on `baz` from the `foo/bar` GitHub identity.
+- A Sigstore client library capable of verifying Sigstore signatures.
+
+Client-side verification of bottles requires `brew` to know which bottles are actually signed. Clients should use the "flag day" metadata above to establish this: `brew install` should refuse to install any unsigned bottle from a tap that both had a flag day _and_ whose flag day is before the local system time. 
+
+The high-level flow for bottle signature verification, at tap time:
+
+- A user performs either an explicit tap (`brew tap foo/bar`) or an implicit tap (`brew install foo/bar/baz`), and the local `brew` client establishes a trust proposition based on that action ("bottles from `foo/bar` must be signed by the `foo/bar` identity").
+- The local `brew` client looks for the "flag day" metadata and, if present and before the local system time, enforces signature verification for all bottles referenced by the tap.
+
+And at `brew install` time:
+
+- For each bottle resolved during a `brew install` invocation, the bottle must have an accompanying signature. If the accompanying signature is not present, then `brew install` must fail.
+- For each bottle and signature, `brew install` must internally reconstruct the _attestation_ described in the signing process above. This is then used as the cleartext input for signature verification. This input is then passed into the underlying Sigstore client along with its associated signature; the Sigstore client is responsible for performing both the signature verification and its associated identity verification.
+- If the underlying signature verification fails, then `brew install` must fail.
+
+Like with previous Homebrew features, an experimental feature-flagged variant of signature verification may be merged and enabled before it is enabled by default.
```

Comment:
For enabled by default, do we mean that if the flag day is set, then the client will always try to verify the signature? Or is the goal to require that all core bottles are signed?

Reply:
> Or is the goal to require that all core bottles are signed?

Yep, that's what I was thinking: my phrasing is a little messy here, but what I had in mind was "verification is optional while we incrementally update/rebuild all bottles with signatures, and once every core bottled is signed then everything *must* be signed after the flag day."

Reply:
If backfill attestations are provided for all the bottles would a flag day be required?  E.g. if all bottles have some attestation then the next release of the client can just assume they'll always be there?

Reply:
> Presumably we can add the flag day logic to clients first, then spend time working on producing signing/attestations, and thus cover a larger group of in-use client versions more quickly when flag day comes.

That's what I was thinking, although @TomHennen's point is also good -- Homebrew tends to "roll" updates very rapidly (i.e., all users rapidly update to the latest stable once released), so with the "backfill" approach there's IMO less value to the flag day.

That being said, I initially worked that in _in case_ we do end up needing it, so IMO it's worth keeping in the proposal as a compatibility/rollout risk mitigation. But we may end up not needing it!

Reply:
Resolved per ref'd commit above.